### DISCUSSION TITLE
Hybrid search on arrays/collections

### DISCUSSION THREAD

**Discussion by nairajay2k** [milvus-io/milvus#31157] [Q&A and General discussion]
How do we do a hybrid search on a collection ona field that is say an array of int or str. Also how do we index such a field? Pl provide some code snippets. I am a milvus newbie. I have done this on qdrant but want to try on milvus.

**Comment by xiaofan-luan** [comment]
hi @nairajay2k 
wanna to know more details about your use cases.

1. most cases you didn't need to built index, filtering on milvus should be extremely fast especially on numeric numbers. (Try not do that on dynamic field)
2. yes we can build index on scalar field https://milvus.io/docs/build_scalar_index.md
3. we are introducing new tantivy index into Milvus, that could be improve filtering speed

**Reply by xiaofan-luan** [reply]
The target release date of tantivy inverted index will be 2.4

**Reply by nairajay2k** [reply]
Hi So I want to have something like an Access Control List  which I plan to have as a seperate column or field in my record. Now each record can have multiple user_ids(int) with access to it. So thats why I thought I would implement this as an array field.

the examples provided are for a single valued scalar field. My use case is that of an array . 
Please provide some code examples for this.

**Reply by xiaofan-luan** [reply]
currently there is no index on array.
you could directly use in expression or contains expression to search.
This feature (array, json and set/map) index will be supported in this June

**Comment by nairajay2k** [comment]
Please give me some sample code

**Comment by xiaofan-luan** [comment]
you just need to use expression

fieldName not in [1, 2, 3]

see this document https://milvus.io/docs/boolean.md

**Reply by nairajay2k** [reply]
I want the reverse...x not in field ...where field is array. Can it be done? Is an array type of field possible in milvus??

**Comment by yhmo** [comment]
@nairajay2k 

This is an example script for searching with array contains or array not contains:

```
import random

from pymilvus import (
    connections,
    FieldSchema, CollectionSchema, DataType,
    Collection,
    utility,
)


connections.connect(host='localhost', port='19530')
print(utility.get_server_version())

collection_name = "test"
dim = 128
metric_type = "L2"

fields=[
    FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=False),
    FieldSchema(name="vector", dtype = DataType.FLOAT_VECTOR, dim=dim),
    FieldSchema(name="int_arr", dtype = DataType.ARRAY, max_capacity=100, element_type=DataType.INT32),
]

schema = CollectionSchema(fields=fields, enable_dynamic_field=False)

if utility.has_collection(collection_name):
    utility.drop_collection(collection_name)

collection = Collection(name=collection_name, schema=schema)
print(f"Collection '{collection_name}' created")
print(collection.describe())

index_params = {
    'metric_type': metric_type,
    'index_type': "FLAT",
    'params': {},
}
collection.create_index(field_name="vector", index_params=index_params)

collection.load()

collection.insert(data={"id": 1, "vector": [random.random() for _ in range(dim)], "int_arr": [1, 2, 3, 4]})
collection.insert(data={"id": 2, "vector": [random.random() for _ in range(dim)], "int_arr": [5, 6]})
collection.insert(data={"id": 3, "vector": [random.random() for _ in range(dim)], "int_arr": [100, 200, 300, 400, 500, 600]})
collection.insert(data={"id": 4, "vector": [random.random() for _ in range(dim)], "int_arr": [22, 24, 26]})
collection.insert(data={"id": 5, "vector": [random.random() for _ in range(dim)], "int_arr": [1, 2]})


def search(expr:str):
    vector = [random.random() for _ in range(dim)]
    params = {
        'metric_type': metric_type,
        "params": {
        }
    }
    results = collection.search(
        data=[vector],
        expr = expr,
        anns_field="vector",
        param=params,
        limit=10,
        output_fields=["int_arr"],
        consistency_level="Strong",
    )

    print(f"\n=============== Search Results with expr '{expr}' ===================")
    for hits in results:
        print(f"topk: {len(hits)}")
        for hit in hits:
            print(hit.to_dict())


search("array_contains(int_arr, 2)")
search("not array_contains(int_arr, 2)")
```


I recommend installing milvus v2.3.x to run this script.
You will get reslut like this:
```
Collection 'test' created
{'collection_name': 'test', 'auto_id': False, 'num_shards': 1, 'description': '', 'fields': [{'field_id': 100, 'name': 'id', 'description': '', 'type': 5, 'params': {}, 'element_type': 0, 'is_primary': True}, {'field_id': 101, 'name': 'vector', 'description': '', 'type': 101, 'params': {'dim': 128}, 'element_type': 0}, {'field_id': 102, 'name': 'int_arr', 'description': '', 'type': 22, 'params': {'max_capacity': '100'}, 'element_type': 4}], 'aliases': [], 'collection_id': 448298418683391749, 'consistency_level': 2, 'properties': {}, 'num_partitions': 1}

=============== Search Results with expr 'array_contains(int_arr, 2)' ===================
topk: 2
{'id': 1, 'distance': 19.36527442932129, 'entity': {'int_arr': [1, 2, 3, 4]}}
{'id': 5, 'distance': 23.167095184326172, 'entity': {'int_arr': [1, 2]}}

=============== Search Results with expr 'not array_contains(int_arr, 2)' ===================
topk: 3
{'id': 3, 'distance': 20.027202606201172, 'entity': {'int_arr': [100, 200, 300, 400, 500, 600]}}
{'id': 4, 'distance': 21.489505767822266, 'entity': {'int_arr': [22, 24, 26]}}
{'id': 2, 'distance': 26.126689910888672, 'entity': {'int_arr': [5, 6]}}

```

**Comment by nairajay2k** [comment]
Thank you so much

On Mon, Mar 11, 2024 at 7:54 AM groot ***@***.***> wrote:

> @nairajay2k <https://github.com/nairajay2k>
>
> This is an example script for searching with array contains or array not
> contains:
>
> import random
>
> from pymilvus import (
>     connections,
>     FieldSchema, CollectionSchema, DataType,
>     Collection,
>     utility,
> )
>
>
> connections.connect(host='localhost', port='19530')
> print(utility.get_server_version())
>
> collection_name = "test"
> dim = 128
> metric_type = "L2"
>
> fields=[
>     FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=False),
>     FieldSchema(name="vector", dtype = DataType.FLOAT_VECTOR, dim=dim),
>     FieldSchema(name="int_arr", dtype = DataType.ARRAY, max_capacity=100, element_type=DataType.INT32),
> ]
>
> schema = CollectionSchema(fields=fields, enable_dynamic_field=False)
>
> if utility.has_collection(collection_name):
>     utility.drop_collection(collection_name)
>
> collection = Collection(name=collection_name, schema=schema)
> print(f"Collection '{collection_name}' created")
> print(collection.describe())
>
> index_params = {
>     'metric_type': metric_type,
>     'index_type': "FLAT",
>     'params': {},
> }
> collection.create_index(field_name="vector", index_params=index_params)
>
> collection.load()
>
> collection.insert(data={"id": 1, "vector": [random.random() for _ in range(dim)], "int_arr": [1, 2, 3, 4]})
> collection.insert(data={"id": 2, "vector": [random.random() for _ in range(dim)], "int_arr": [5, 6]})
> collection.insert(data={"id": 3, "vector": [random.random() for _ in range(dim)], "int_arr": [100, 200, 300, 400, 500, 600]})
> collection.insert(data={"id": 4, "vector": [random.random() for _ in range(dim)], "int_arr": [22, 24, 26]})
> collection.insert(data={"id": 5, "vector": [random.random() for _ in range(dim)], "int_arr": [1, 2]})
>
>
> def search(expr:str):
>     vector = [random.random() for _ in range(dim)]
>     params = {
>         'metric_type': metric_type,
>         "params": {
>         }
>     }
>     results = collection.search(
>         data=[vector],
>         expr = expr,
>         anns_field="vector",
>         param=params,
>         limit=10,
>         output_fields=["int_arr"],
>         consistency_level="Strong",
>     )
>
>     print(f"\n=============== Search Results with expr '{expr}' ===================")
>     for hits in results:
>         print(f"topk: {len(hits)}")
>         for hit in hits:
>             print(hit.to_dict())
>
>
> search("array_contains(int_arr, 2)")
> search("not array_contains(int_arr, 2)")
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/milvus-io/milvus/discussions/31157#discussioncomment-8739724>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJ6XWJ425VC65CKJWYRFQ5LYXUIXXAVCNFSM6AAAAABEOGO4CGVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DOMZZG4ZDI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Regards
Ajay
8762803277
